parasol_cpu/src/proc/ops/load.rs

```diff
@@ -1,91 +1,98 @@
     /// Execute a load instruction.
     pub fn load(
         &mut self,
         retirement_info: RetirementInfo<DispatchIsaOp>,
-        scoreboard_entry: ScoreboardEntryRef<DispatchIsaOp>,
-        src: RobEntryRef<PtrRegister>,
+        memory: &Memory,
+        src: RobEntryRef<Register>,
         dst: RobEntryRef<Register>,
         width: u32,
         instruction_id: usize,
-        pc: usize,
+        pc: u32,
     ) {
-        let load_impl = |scoreboard_entry: &ScoreboardEntryRef<DispatchIsaOp>| -> Result<()> {
+        let load_impl = || -> Result<()> {
             unwrap_registers!((mut dst) (src));
 
-            let num_bytes = width.next_multiple_of(8) as usize / 8;
-            let mask = read_write_mask(width);
-
             match src {
-                PtrRegister::Plaintext(ptr) => {
-                    check_offset(width, ptr.offset, ptr.base.len(), instruction_id, pc)?;
-
-                    // Plaintext loads and stores exec immediately, so we don't need to update our deps.
-
-                    let mut result = 0u128;
-
-                    for i in 0..(num_bytes - 1) {
-                        result |= (*ptr.base[ptr.offset as usize + i].borrow() as u128) << (8 * i);
+                Register::Plaintext { val: ptr, width: _ } => {
+                    let base_addr = *ptr as u32;
+
+                    if base_addr % width != 0
+                        || width > 16
+                        || !width.is_power_of_two()
+                        || width == 0
+                    {
+                        return Err(Error::UnalignedAccess(base_addr));
                     }
 
-                    result |= ((mask & *ptr.base[ptr.offset as usize + num_bytes - 1].borrow())
-                        as u128)
-                        << (8 * (num_bytes - 1));
-
-                    *dst = Register::Plaintext { val: result, width };
+                    let base_addr = Ptr32::from(base_addr);
 
-                    FheProcessor::retire(&retirement_info, Ok(()));
-                }
-                PtrRegister::Ciphertext(CiphertextPtr::PlainOffset(ptr)) => {
-                    check_offset(width, ptr.offset, ptr.base.len(), instruction_id, pc)?;
-                    ptr.on_read(scoreboard_entry);
-
-                    let result = (0..width)
-                        .map(|_| Arc::new(AtomicRefCell::new(self.aux_data.enc.allocate_glwe_l1())))
-                        .collect::<Vec<_>>();
-
-                    let range = ptr.offset as usize * 8..ptr.offset as usize * 8 + width as usize;
-
-                    for (input, output) in ptr.base[range].iter().zip(result.iter()) {
-                        let input = AtomicRefCell::borrow(input);
-                        let mut output = AtomicRefCell::borrow_mut(output);
-
-                        output.clone_from(&input);
-                    }
+                    // Plaintext loads and stores exec immediately, so we don't need to update our deps.
```

Comment:
Is the word "Plaintext" valid in this comment? The ciphertext case also seems to execute immediately.